xc/domain.ml

```diff
@@ -633,6 +633,49 @@ let create_channels ~xc uuid domid =
   debug "VM = %s; domid = %d; store evtchn = %d; console evtchn = %d" (Uuid.to_string uuid) domid store console;
   store, console
 
+let numa_hierarchy =
+  let open Xenctrlext in
+  let open Topology in
+  Lazy.from_fun (fun () ->
+      let distances = Xenctrlext.(with_xc numainfo).distances in
+      let cpu_to_node = Xenctrlext.(with_xc cputopoinfo) |> Array.map (fun t -> t.node) in
+      NUMA.v ~distances ~cpu_to_node)
+
+let numa_mutex = Mutex.create ()
+let numa_resources = ref None
+
+let numa_placement domid ~vcpus ~memory =
+  let open Xenctrlext in
+  let open Topology in
+  let hint = Mutex.execute numa_mutex (fun () ->
+      let host = Lazy.force numa_hierarchy in
```

Comment:
why not

```
let numa_hierarchy () =
  ...
  NUMA.v ~distances ~cpu_to_node

...
let host = numa_hierarchy () in
...
```
?

Reply:
I want to evaluate it only once, in case it is costly to build the initial node candidate list.